Add instructions for building and testing with Rust (#35725)
github/docs: content/actions/use-cases-and-examples/building-and-testing/building-and-testing-rust.md

## Using a Rust workflow template

{% data reusables.actions.workflow-templates-get-started %}

{% data variables.product.prodname_dotcom %} provides a Rust workflow template that should work for most basic Rust projects. The subsequent sections of this guide give examples of how you can customize this workflow template.

{% data reusables.repositories.navigate-to-repo %}
{% data reusables.repositories.actions-tab %}
{% data reusables.actions.new-starter-workflow %}
1. The "Choose a workflow" page shows a selection of recommended workflow templates. Search for "Rust".
1. Filter the selection of workflows by clicking **Continuous integration**.
1. On the "Rust - by {% data variables.product.prodname_actions %}" workflow, click **Configure**.

   ![Screenshot of the "Choose a workflow" page. The "Configure" button on the "Rust" workflow is highlighted with an orange outline.](/assets/images/help/actions/starter-workflow-rust.png)
1. Edit the workflow as required. For example, change the version of Rust.
1. Click **Commit changes**.

> {% ifversion fpt or ghec %}

Reviewer: Right, I did look into what's hidden behind the conditional, and didn't really get it — it only hides where the workflow is being committed. Can't tell why that's only relevant for fpt & ghec, but it's cargo-culted from other articles so there's not specific versioning needed here — but perhaps check generally if this info needs to be hidden from some versions.

> {% ifversion fpt or ghec %}

Reviewer 2: Thanks. This looks correct to me ✨ 

I've added the GitHub Enterprise Server-specific information earlier in the file. So this article should apply to all versions.

```

Note that the `release` keyword used above, corresponds to a cargo profile. You can use any [profile](https://doc.rust-lang.org/cargo/reference/profiles.html) you have defined in your `Cargo.toml` file.

## Uploading and publishing artifacts

In case publishing artifacts is needed, but not to crates.io, the following example demonstrates how to upload artifacts to the workflow run:

```yaml copy
      - name: Upload hello app
        uses: {% data reusables.actions.action-upload-artifact %}
        with:
          name: cndk8-hello
          path: target/${{ matrix.BUILD_TARGET }}/cndk8
```

And to use them on a different job, i.e publishing as a github release, check [AUTOTITLE](/actions/security-for-github-actions/security-guides/automatic-token-authentication) to ensure your workflows have the right permissions on the repository.

```yaml copy
      - uses: {% data reusables.actions.action-checkout %}
      - name: Download hello app
        uses: {% data reusables.actions.action-download-artifact %}
        with:
          name: cndk8-hello
          path: ./cndk8-hello
      - name: Publish hello app binary as a release on GitHub
      - run: |

>           gh release create v0.1.0 --generate-notes

Reviewer: (TBH I'd avoid creating the release in CI at all; and instead publish the artifacts on release trigger — attaching the assets to the recent/created release instead?)

>           gh release create v0.1.0 --generate-notes

Reviewer 2: Based on https://cli.github.com/manual/gh_release_create, I'm wondering if we could replace both lines with:

```suggestion
          gh release create --generate-notes ./{% raw %}<my-app>/<my-project>#<my-app>{% endraw %}
```

>           gh release create v0.1.0 --generate-notes

Reviewer 3: @felicitymay I think the replacement is the way to go.

          gh release upload v0.1.0 ./cndk8-hello/cndk8#my-cndk8-hello-app
```

## Publishing your package or library to crates.io

Once you have setup your workflow to build and test your code, you can alternatively use a secret to login to crates.io and publish your package.

```yaml copy
      - name: login into crates.io
        run: cargo login ${{ secrets.CRATES_IO }}
      - name: Build binaries in "release" mode
        run: cargo build -r
      - name: "Package for crates.io"
        run: cargo package # publishes a package as a tarball
      - name: "Publish to crates.io"
        run: cargo publish # publishes your crate as a library that can be added as a dependency
```

> As an example of how packages are published, see the [cndk8 0.1.0](https://crates.io/crates/cndk8/0.1.0). In the case that there are errors with Metadata check

Reviewer: e.g. in the case of .NET that needs an argument with some placeholder filled in, it uses the GH Docs convention of using angle brackets:

`run: dotnet pack --configuration Release <my project>`

The examples here should too make it obvious what are just placeholder values in need of filling in correctly before the examples start to work for the reader…

      - uses: {% data reusables.actions.action-checkout %}
      - name: Build binaries in "${{ matrix.BUILD_TARGET }}" mode
        run: cargo build --profile ${{ matrix.BUILD_TARGET }}
      - name: Run tests in "${{ matrix.BUILD_TARGET }}" mode
        run: cargo test --profile ${{ matrix.BUILD_TARGET }}
```

Note that the `release` keyword used above, corresponds to a cargo profile. You can use any [profile](https://doc.rust-lang.org/cargo/reference/profiles.html) you have defined in your `Cargo.toml` file.

## Upload artifacts

In case publishing artifacts is needed, but not to crates.io, the following example demonstrates how to upload artifacts to the workflow run:

```yaml copy
      - name: Upload hello app
        uses: {% data reusables.actions.action-upload-artifact %}
        with:
          name: cndk8-hello
          path: target/${{ matrix.BUILD_TARGET }}/cndk8
```

And to use them on a different job, i.e publishing:

```yaml copy
      - name: Download hello app
        uses: {% data reusables.actions.action-download-artifact %}
        with:
          name: cndk8-hello
          path: ./cndk8-hello
      - name: Publish hello app to GitHub Packages
        run: |

>           curl -u "${{ github.actor }}:${{ secrets.GH_TOKEN }}" \

Reviewer 3: Should this be `GITHUB_TOKEN` here, rather than `GH_TOKEN`?
```suggestion
          curl -u "${{ github.actor }}:${{ secrets.GITHUB_TOKEN }}" \
```

            -X POST "https://uploads.github.com/repos/${{ github.repository }}/releases/assets?name=cndk8-hello.tar.gz" \
            --header "Content-Type: application/gzip" \

>             --data-binary @./cndk8-hello/cndk8

Author: > What do you think about switching this to use something like gh release upload instead?
@mattr- I feel we could remove this all together; adding a step that does `gh release upload`, goes more into the question I had already in [this comment](https://github.com/github/docs/pull/35725#discussion_r1901166219) of whether there would/could be a github registry for rust packages... 

> haven't been able to find documentation about uploads.github.com

>             --data-binary @./cndk8-hello/cndk8

Author: it was actually easier than I thought: https://github.com/foursixnine/hetzner-robot-rs/releases/tag/v0.1.0 was published using the new steps described in this guide: https://github.com/foursixnine/hetzner-robot-rs/actions/runs/12828374178/job/35772361174#step:4:4

---
title: Building and testing Rust
intro: You can create a continuous integration (CI) workflow to build and test your Rust project.
versions:
  ghec: '*'
type: tutorial
topics:
  - CI
shortTitle: Build, test & Publish with Rust
---

> ~~{% data reusables.actions.enterprise-github-hosted-runners %}~~

Reviewer 2: This is causing the formatting error (tested locally). Suggest changing into an HTML comment instead. 

```suggestion
<!--{% data reusables.actions.enterprise-github-hosted-runners %}-->
```

Author: thanks! indeed that fixes the formatting issues

{% data reusables.repositories.actions-tab %}
{% data reusables.actions.new-starter-workflow %}
1. The "Choose a workflow" page shows a selection of recommended workflow templates. Search for "Rust".
1. Filter the selection of workflows by clicking **Continuous integration**.
1. On the "Rust - by {% data variables.product.prodname_actions %}" workflow, click **Configure**.

   ![Screenshot of the "Choose a workflow" page. The "Configure" button on the "Rust" workflow is highlighted with an orange outline.](/assets/images/help/actions/starter-workflow-rust.png)
1. Edit the workflow as required. For example, change the version of Rust.
1. Click **Commit changes**.

{% ifversion fpt or ghec %}
   The `rust.yml` workflow file is added to the `.github/workflows` directory of your repository.
{% endif %}

## Specifying a Rust version

At the time of writing, the default rust compiler version is 1.83.0 rustup is available and can be used to install additional toolchains.

```yaml copy
      - name: Temporarily modify the rust toolchain version
        run: rustup override set nightly
      - name: Output rust version for educational purposes
        run: rustup --version

> ```

Reviewer: Before any other cargo steps, to set up the runner env / toolchain versions as a first step.

> ```

Reviewer: (i.e. the same place you'd use `setup-python` in a python workflow… if you need changing versions/stability)

        run: cargo test --profile ${{ matrix.BUILD_TARGET }}
```

Note that the `release` keyword used above, corresponds to a cargo profile. You can use any [profile](https://doc.rust-lang.org/cargo/reference/profiles.html) you have defined in your `Cargo.toml` file.

## Upload artifacts

In case publishing artifacts is needed, but not to crates.io, the following example demonstrates how to upload artifacts to the workflow run:

```yaml copy
      - name: Upload hello app
        uses: {% data reusables.actions.action-upload-artifact %}
        with:
          name: cndk8-hello
          path: target/${{ matrix.BUILD_TARGET }}/cndk8
```

And to use them on a different job, i.e publishing:

```yaml copy
      - name: Download hello app
        uses: {% data reusables.actions.action-download-artifact %}
        with:
          name: cndk8-hello
          path: ./cndk8-hello
      - name: Publish hello app to GitHub Packages
        run: |
          curl -u "${{ github.actor }}:${{ secrets.GH_TOKEN }}" \
            -X POST "https://uploads.github.com/repos/${{ github.repository }}/releases/assets?name=cndk8-hello.tar.gz" \
            --header "Content-Type: application/gzip" \

>             --data-binary @./cndk8-hello/cndk8

Author: I wonder if this is wanted/desired or if gh has plans to offer a rust registry

1. Edit the workflow as required. For example, change the version of Rust.
1. Click **Commit changes**.

{% ifversion fpt or ghec %}
   The `rust.yml` workflow file is added to the `.github/workflows` directory of your repository.
{% endif %}

## Specifying a Rust version

At the time of writing, the default rust compiler version is 1.83.0 rustup is available and can be used to install additional toolchains. For example, the following workflow temporarily sets the toolchain to nightly:

```yaml copy
      - name: Temporarily modify the rust toolchain version
        run: rustup override set nightly
      - name: Ouput rust version for educational purposes
        run: rustup --version
```

### Caching dependencies

You can cache and restore dependencies using the following example below. Note that you will need to have Cargo.lock in your repository to cache dependencies.

```yaml copy
      - name: ⚡ Cache
        uses: actions/cache@v4
        with:
          path: |
            ~/.cargo/registry
            ~/.cargo/git
            target

>           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}

Reviewer 2: The `{{ }}` in this line seem to be causing the linter failure. We can avoid this by disabling Liquid processing for the line. See `{% raw %} in [Code blocks](https://docs.github.com/en/contributing/style-guide-and-content-model/style-guide#code-blocks).

```suggestion
          key: {% raw %}${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}{% endraw %}
```

The [error rending the new page](https://github.com/github/docs/actions/runs/12573366781/job/35046299814?pr=35725#step:5:21) might well be why it's not visible in the preview.

Author: Thanks a lot! that "fixed it", there seem to be other [problems](https://docs-35725-22a95e.preview.ghdocs.com/en/enterprise-cloud@latest/actions/use-cases-and-examples/building-and-testing/building-and-testing-rust): I'll look into that later

## Prerequisites

You should already be familiar with YAML syntax and how it's used with {% data variables.product.prodname_actions %}. For more information, see [AUTOTITLE](/actions/using-workflows/workflow-syntax-for-github-actions).

We recommend that you have a basic understanding of the Rust language. For more information, see [Getting started with Rust](https://www.rust-lang.org/learn).

## Using a Rust workflow template

{% data reusables.actions.workflow-templates-get-started %}

{% data variables.product.prodname_dotcom %} provides a Rust workflow template that should work for most basic Rust projects. The subsequent sections of this guide give examples of how you can customize this workflow template.

{% data reusables.repositories.navigate-to-repo %}
{% data reusables.repositories.actions-tab %}
{% data reusables.actions.new-starter-workflow %}
1. The "Choose a workflow" page shows a selection of recommended workflow templates. Search for "Rust".
1. Filter the selection of workflows by clicking **Continuous integration**.
1. On the "Rust - by {% data variables.product.prodname_actions %}" workflow, click **Configure**.

>    ![Screenshot of the "Choose a workflow" page. The "Configure" button on the "Rust" workflow is highlighted with an orange outline.](/assets/images/help/actions/starter-workflow-rust.png)

Reviewer 2: I'm not sure why this image is shown as broken in the preview. It works fine locally. I'm guessing this is a bug in our new preview system.

Reviewer 2: Asking internally, I found that images are intentionally omitted from the preview. This works fine for me locally

```yaml copy
      - name: Upload release artifact
        uses: {% data reusables.actions.action-upload-artifact %}
        with:
          name: {% raw %}<my-app>{% endraw %}
          path: {% raw %}target/${{ matrix.BUILD_TARGET }}/<my-app>{% endraw %}
```

To use the uploaded artifact in a different job, ensure your workflows have the right permissions for the repository, see [AUTOTITLE](/actions/security-for-github-actions/security-guides/automatic-token-authentication). You could use these example steps to download the app created in the previous workflow and publish it on {% data variables.product.github %}.

```yaml copy
      - uses: {% data reusables.actions.action-checkout %}
      - name: Download release artifact
        uses: {% data reusables.actions.action-download-artifact %}
        with:
          name: {% raw %}<my-app>{% endraw %}
          path: ./{% raw %}<my-app>{% endraw %}
      - name: Publish built binary to {% data variables.product.github %} releases
      - run: |
          gh release create v0.1.0 --generate-notes

>           gh release upload v0.1.0 ./{% raw %}<my-app>/<my-project>#<my-app>{% endraw %}
> ```

Reviewer 2: ```suggestion
```
```

[---
title: Building and testing Rust
intro: You can create a continuous integration (CI) workflow to build and test your Rust project.
versions:
  fpt: '*'
  ghes: '*'

>   ghec: '*'

Author: what would be the correct options then? I don't know ftp, ghes or ghec... what would you suggest?

type: tutorial
topics:
  - CI
shortTitle: Build & test Rust
redirect_from:
  - /actions/automating-builds-and-tests/building-and-testing-rust
---

~~{% data reusables.actions.enterprise-github-hosted-runners %}~~

## Introduction

> This guide shows you how to build, test, and publish a Rust package.

Author: I'll add the instructions for publishing to crates.io, as I don't know if GitHub has one for rust specifically..

https://github.com/foursixnine/cndk8/pull/3

{% data variables.product.prodname_dotcom %}-hosted runners have a tools cache with preinstalled software, which includes the dependencies for Go. For a full list of up-to-date software and the preinstalled versions of Go, see [AUTOTITLE](/actions/using-github-hosted-runners/about-github-hosted-runners#preinstalled-software).

## Prerequisites

You should already be familiar with YAML syntax and how it's used with {% data variables.product.prodname_actions %}. For more information, see [AUTOTITLE](/actions/using-workflows/workflow-syntax-for-github-actions).

We recommend that you have a basic understanding of the Rust language. For more information, see [Getting started with Rust](https://www.rust-lang.org/learn).

## Using a Rust workflow template

{% data reusables.actions.workflow-templates-get-started %}

> {% data variables.product.prodname_dotcom %} provides a Rust workflow template that should work for most basic Rust projects. The subsequent sections of this guide give examples of how you can customize this workflow template.

Author: @janbrasna yep

{% data reusables.repositories.navigate-to-repo %}
{% data reusables.repositories.actions-tab %}
{% data reusables.actions.new-starter-workflow %}
1. The "Choose a workflow" page shows a selection of recommended workflow templates. Search for "Rust".
1. Filter the selection of workflows by clicking **Continuous integration**.
1. On the "Rust - by {% data variables.product.prodname_actions %}" workflow, click **Configure**.

>    ![Screenshot of the "Choose a workflow" page. The "Configure" button on the "Go" workflow is highlighted with an orange outline.](/assets/images/help/actions/starter-workflow-go.png)

Author: image added in 7a2354f3c399183cd10dc7bb43aa0b0047787b8a

1. On the "Rust - by {% data variables.product.prodname_actions %}" workflow, click **Configure**.

   ![Screenshot of the "Choose a workflow" page. The "Configure" button on the "Rust" workflow is highlighted with an orange outline.](/assets/images/help/actions/starter-workflow-rust.png)

{%- ifversion ghes %}
   If you don't find the "Rust - by {% data variables.product.prodname_actions %}" workflow template, copy the following workflow code to a new file called `rust.yml` in the `.github/workflows` directory of your repository.

   ```yaml copy
   name: Rust

   on:
     push:
       branches: [ "main" ]
     pull_request:
       branches: [ "main" ]

   env:

>      CARGO_TERM_COLOR: always

Reviewer 3: Does Actions filter out terminal escape codes now? The safest option is to set this to `never`
```suggestion
     CARGO_TERM_COLOR: never
```

1. On the "Rust - by {% data variables.product.prodname_actions %}" workflow, click **Configure**.

   ![Screenshot of the "Choose a workflow" page. The "Configure" button on the "Rust" workflow is highlighted with an orange outline.](/assets/images/help/actions/starter-workflow-rust.png)

> {%- ifversion ghes %}

Reviewer 2: This section is for GitHub Enterprise Server users who may not have access to GitHub Marketplace.

## Specifying a Rust version

> {% data variables.product.prodname_dotcom %}-hosted runners include a recent version of the rustup rust compiler. You can use rustup to report on the version installed on a runner, override the version, and to install additional toolchains. For more information, see [The rustup book](https://rust-lang.github.io/rustup/).

Reviewer 2: Revised to remove the reference to an explicit version of rustup to stop the information going out of date.

> {% data variables.product.prodname_dotcom %}-hosted runners include a recent version of the rustup rust compiler. You can use rustup to report on the version installed on a runner, override the version, and to install additional toolchains. For more information, see [The rustup book](https://rust-lang.github.io/rustup/).

Reviewer: This may need a xcheck from some rust-universe SME, but I'd phrase it this way:

```suggestion
{% data variables.product.prodname_dotcom %}-hosted runners include a recent version of the Rust toolchain. You can use rustup to report on the version installed on a runner, override the version, and to install different toolchains. For more information, see [The rustup book](https://rust-lang.github.io/rustup/).
```

      matrix:
        BUILD_TARGET: [release] # refers to a cargo profile
    outputs:
      release_built: {% raw %}${{ steps.set-output.outputs.release_built }}{% endraw %}
    steps:
      - uses: {% data reusables.actions.action-checkout %}
      - name: Build binaries in "{% raw %}${{ matrix.BUILD_TARGET }}{% endraw %}" mode
        run: cargo build --profile ${% raw %}{{ matrix.BUILD_TARGET }}{% endraw %}
      - name: Run tests in "${% raw %}{{ matrix.BUILD_TARGET }}{% endraw %}" mode
        run: cargo test --profile ${% raw %}{{ matrix.BUILD_TARGET }}{% endraw %}
```

The `release` keyword used in this example corresponds to a cargo profile. You can use any [profile](https://doc.rust-lang.org/cargo/reference/profiles.html) you have defined in your `Cargo.toml` file.

> ## Publishing your package or library to crates.io

Reviewer 2: I'm assuming that this is the most common use case, so I moved it so that it was before the section on uploading artifacts to GitHub.

```yaml copy
      - name: Login into crates.io
        run: cargo login {% raw %}${{ secrets.CRATES_IO }}{% endraw %}
      - name: Build binaries in "release" mode
        run: cargo build -r
      - name: "Package for crates.io"
        run: cargo package # publishes a package as a tarball
      - name: "Publish to crates.io"
        run: cargo publish # publishes your crate as a library that can be added as a dependency
```

If there are any errors building and packaging the crate, check the metadata in your manifest, `Cargo.toml` file, see [The Manifest Format](https://doc.rust-lang.org/cargo/reference/manifest.html). You should also check your `Cargo.lock` file, see [Cargo.toml vs Cargo.lock](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html).

## Packaging workflow data as artifacts

After a workflow completes, you can upload the resulting artifacts for analysis or to use in another workflow. You could add these example steps to the workflow to upload an application for use by another workflow.

```yaml copy

>       - name: Upload {% raw %}<my-app>{% endraw %}

Reviewer: ```suggestion
      - name: Upload release artifact
```

        uses: {% data reusables.actions.action-upload-artifact %}
        with:
          name: {% raw %}<my-app>{% endraw %}
          path: {% raw %}target/${{ matrix.BUILD_TARGET }}/<my-app>{% endraw %}
```

To use the uploaded artifact in a different job, ensure your workflows have the right permissions for the repository, see [AUTOTITLE](/actions/security-for-github-actions/security-guides/automatic-token-authentication). You could use these example steps to download the app created in the previous workflow and publish it on {% data variables.product.github %}.

```yaml copy
      - uses: {% data reusables.actions.action-checkout %}

>       - name: Download {% raw %}<my-app>{% endraw %}

Reviewer: ```suggestion
      - name: Download release artifact
```

        uses: {% data reusables.actions.action-download-artifact %}
        with:
          name: {% raw %}<my-app>{% endraw %}
          path: ./{% raw %}<my-app>{% endraw %}

>       - name: Publish {% raw %}<my-app>{% endraw %} binary as a release on {% data variables.product.github %}

Reviewer: ```suggestion
      - name: Publish built binary to {% data variables.product.github %} releases
```
